qiskit/circuit/register.py

```diff
@@ -59,46 +57,40 @@ def __init__(self, size, name=None):
             if name_format.match(name) is None:
                 raise QiskitError("%s is an invalid OPENQASM register name." % name)
 
-        self.name = name
-        self.size = size
+        self._name = name
+        self._hash = None
+        for item in range(size):
+            self.append(self.bit_type(self, item))
 
     def __repr__(self):
         """Return the official string representing the register."""
-        return "%s(%d, '%s')" % (self.__class__.__qualname__, self.size, self.name)
+        return "%s(%d, '%s')" % (self.__class__.__qualname__, len(self), self.name)
 
-    def __len__(self):
-        """Return register size"""
-        return self.size
+    @property
+    def name(self):
+        """Returns the name of the register"""
+        if self._name is None:
+            self._name = '%s%i' % (self.prefix, next(self.instances_counter))
```

Comment:
why is this piece of code moved here? I don't think the property getter should assign a name, it should just return whatever name is there.

Reply:
Because lazy automatic name generation. Instead of creating a name at construction time, we only create a name if that attribute is needed.